### ISSUE TITLE
[URGENT - LIVE VERSION] User has been charged but purchase isn't showing up

### ISSUE THREAD

**Issue by melih-veloxia** [RevenueCat/purchases-unity#90]
- [x] I have updated Purchases SDK to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/RevenueCat/purchases-unity/blob/main/CONTRIBUTING.md)
- [x] I have searched the [Community](https://community.revenuecat.com)
- [x] I have read [docs.revenuecat.com](https://docs.revenuecat.com/)
- [x] I have searched for [existing Github issues](https://github.com/RevenueCat/purchases-unity/issues)

**Describe the bug**
When there is a billing problem with the card or user is asked for a confirmation with a password, the purchase ends up outside of the game. `userCancelled` becomes false but as `STORE_ERROR` is returned, purchase process returns on client. At the end of the process, user gets charged without having the rewards. Not reproducible on sandbox environment.
**Do not remove any of the steps from the template below. If a step is not applicable to your issue, please leave that step empty.**

1. Environment
   1. Platform: iOS/Android
   2. SDK version: 3.5.0
   3. OS + version:
   4. Xcode/Android Studio version:
   5. Unity version: 2020.3.25f1
   6. How widespread is the issue. Percentage of devices affected. ~15%
2. [Debug logs](https://docs.revenuecat.com/docs/debugging) that reproduce the issue
3. Steps to reproduce, with a description of expected vs. actual behavior
Release a bundle to stores
Make a real purchase with a real card without required limit inside the card
Store asks you to fill another payment method
Complete the purchase
4. **Other information** (e.g. stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, etc.)

**Additional context**

**Comment by aboedo** [comment]
hi @melih-veloxia!! 👋 

When a purchase is finished outside of the regular flow (i.e.: if there's a problem with the card, or if the user has to ask for parent's permission for a purchase, etc), the SDK will still provide a callback to the app. 

This is done through the `PurchaserInfoListener`. Here are [our docs explaining how to use it](https://docs.revenuecat.com/docs/purchaserinfo#listening-for-purchaser-info-updates).

Basically in Unity you just make sure that you override the method `PurchaserInfoReceived` in the class that inherits from `Purchases.UpdatedPurchaserInfoListener`, then react to changes when the method gets called. 

Hope this helps!

**Comment by aboedo** [comment]
`purchaserInfo` has information about entitlements, which in turn has information about the products the user has purchased. 

We don't however have a specific way to know exactly _which_ purchase just went through, sadly (this is in our backlog). 

So the method you described (setting up the listener, then when the app is ready to do something with the purchased content, checking the value of `purchaserInfo`) should work.

Are there particular use cases that you're concerned about that wouldn't be covered by this?

**Comment by melih-veloxia** [comment]
@aboedo entitlements returns only subscriptions, not the in-app-purchases. To check the in-app-purchases, I may use `nonSubscriptionTransactions`. To handle these kind of situations perfectly, I need to parse both entitlements and `nonSubscriptionTransactions` into our local database and check the latest ones, match it with the latest failed one, then mark them as rewardGiven to avoid reward duplication. It is such a huge process that I deeply feel like we shouldn't handle it on our own side to make it perfect as RevenueCat is a very well paid tool. 

Please let me know in case there is an easier solution I missed.

**Comment by aboedo** [comment]
@melih-veloxia I agree with you that this is something that RevenueCat could and should handle.

We have it in our backlog, however I don't have an ETA. 

In the meantime, the solution you described is as good as it gets, I'm afraid. 

Hope that makes sense
